overlord/devicestate/devicestate.go

```diff
@@ -475,52 +476,104 @@ func (ms *modelSnapsForRemodel) canHaveUC18PinnedTrack() bool {
+			if errors.Is(err, &snap.NotInstalledError{}) {
+				return nil, fmt.Errorf("internal error: no snap file provided for %q", snapName)
+			}
+			return nil, err
+		}
+
+		if opts != nil && !opts.Revision.Unset() && info.Revision != opts.Revision {
+			return nil, fmt.Errorf("installed snap %q does not match revision required to be used for offline remodel: %s != %s", snapName, opts.Revision, info.Revision)
+		}
+
+		// this would only occur from programmer error, since
+		// UpdateWithDeviceContext should only be called if either the snap
+		// revision or channel needs to change. once we get here, we know that
+		// the revision is the same, so the channel should be different.
+		if opts == nil || opts.Channel == info.Channel {
+			return nil, fmt.Errorf("internal error: installed snap %q already on channel %q", snapName, info.Channel)
+		}
+
+		// since snapstate.Switch doesn't take a prereq tracker, we need to add
+		// it explicitly
+		tracker.Add(info)
+
+		return snapstateSwitch(st, snapName, opts)
 	}
 	return snapstateUpdateWithDeviceContext(st, snapName, opts,
 		userID, snapStateFlags, tracker, deviceCtx, fromChange)
 }
 
 func (ro *remodelVariant) InstallWithDeviceContext(ctx context.Context, st *state.State,
-	pathSI *pathSideInfo, snapName string, opts *snapstate.RevisionOptions, userID int,
+	snapName string, snapID string, opts *snapstate.RevisionOptions, userID int,
 	snapStateFlags snapstate.Flags, tracker snapstate.PrereqTracker, deviceCtx snapstate.DeviceContext,
 	fromChange string) (*state.TaskSet, error) {
 	logger.Debugf("snap %s needs install", snapName)
 	if opts == nil {
 		opts = &snapstate.RevisionOptions{}
 	}
-	if ro.localSnapsRequired {
+	if ro.offline {
+		pathSI := ro.sideInfoAndPathFromID(snapID)
 		if pathSI == nil {
 			return nil, fmt.Errorf("no snap file provided for %q", snapName)
 		}
+
 		return snapstateInstallPathWithDeviceContext(st, pathSI.localSi, pathSI.path, snapName, opts,
 			userID, snapStateFlags, tracker, deviceCtx, fromChange)
 	}
 	return snapstateInstallWithDeviceContext(ctx, st, snapName,
 		opts, userID, snapStateFlags, tracker, deviceCtx, fromChange)
 }
 
+// sideInfoAndPathFromID returns the SideInfo/path for a given snap ID. Note
+// that this will work only for asserted snaps, that is the only case we
+// support for remodeling at the moment.
+func (ro *remodelVariant) sideInfoAndPathFromID(id string) *pathSideInfo {
+	for i, si := range ro.localSnaps {
+		if si.SnapID == id {
+			return &pathSideInfo{localSi: ro.localSnaps[i], path: ro.localSnapPaths[i]}
+		}
+	}
+	// We do not return an error because
+	// 1. We call the function also when there are no local snaps,
+	//    so not finding is expected.
+	// 2. Even if local snaps are required, it is not known yet if
+	//    the one identified by id is really needed (it could have
+	//    been already installed, etc.). So the returned SideInfo is
+	//    checked later in {Install,Update}WithDeviceContext.
```

Comment:
This is not on you (you just moved the function to be a member of remodelVariant), but due to your refactorings here this comment is no longer completely true. Because in the case of `InstallWithDeviceContext` suddenly it is an error to not find it, making this untrue. In addition these are *valuable* comments which I actually were about to ask explanation for up where this function is called in `UpdateWithDeviceContext` (until I by accident saw this since this function was moved). 

Maybe this should be renamed to `maybeSideInfoAndPathFromID` to identify that this function may return nil, and then these comments could be moved up to the usage in `UpdateWithDeviceContext` to explain what we do in the case there is no SideInfo.

Reply:
Good call. I renamed the method and added some clarifying comments.